### PR TITLE
refactor(common): Convert `HttpStatusCode` enum to const object

### PR DESCRIPTION
**Pull request by JeanMeche**
Enums are not inlined, this results in the 3Kb minified output for this single enum. This commit tends to optimise that.

A presubmit will tell us how breaking this change is.

### CODE REVIEW

**Comment by kbrilla** [comment]
in this PR https://github.com/angular/angular/pull/51670#discussion_r1317566943 it is said that esbuild does inline enums

**Comment by JeanMeche** [comment]
@kbrilla It does only when directly fed TS code : [esbuild](https://esbuild.github.io/try/#YgAwLjE5LjIALS1idW5kbGUgLS1zcGxpdHRpbmcgLS1vdXRkaXI9b3V0IC0tZm9ybWF0PWVzbQBlAGVudHJ5LnRzAGltcG9ydCB7SHR0cFN0YXR1c0NvZGV9IGZyb20gJy4vaHR0cC5qcyc7Cgpjb25zb2xlLmxvZyhIdHRwU3RhdHVzQ29kZS5JQW1BVGVhcG90KTsKCihhc3luYyAoKSA9PiB7CiAgaW1wb3J0KCcuL2ZpbGUuanMnKTsKfSkoKTsAAGh0dHAudHMAZXhwb3J0IGVudW0gSHR0cFN0YXR1c0NvZGUgewogIElBbUFUZWFwb3QgPSA0MTgsCn0KCmNvbnNvbGUubG9nKEh0dHBTdGF0dXNDb2RlLklBbUFUZWFwb3QpOwAAZmlsZS5qcwBpbXBvcnQge0h0dHBTdGF0dXNDb2RlfSBmcm9tICcuL2h0dHAuanMnOwoKY29uc29sZS5sb2coSHR0cFN0YXR1c0NvZGUuSUFtQVRlYXBvdCk7)

But if you pass it the TSC output like the CLI does its doesn't : [esbuild](https://esbuild.github.io/try/#YgAwLjE5LjIALS1idW5kbGUgLS1zcGxpdHRpbmcgLS1vdXRkaXI9b3V0IC0tZm9ybWF0PWVzbQBlAGVudHJ5LnRzAGltcG9ydCB7SHR0cFN0YXR1c0NvZGV9IGZyb20gJy4vaHR0cC5qcyc7Cgpjb25zb2xlLmxvZyhIdHRwU3RhdHVzQ29kZS5JQW1BVGVhcG90KTsKCihhc3luYyAoKSA9PiB7CiAgaW1wb3J0KCcuL2ZpbGUuanMnKTsKfSkoKTsAAGh0dHAudHMAZXhwb3J0IHZhciBIdHRwU3RhdHVzQ29kZTsKKGZ1bmN0aW9uIChIdHRwU3RhdHVzQ29kZSkgewogICAgSHR0cFN0YXR1c0NvZGVbSHR0cFN0YXR1c0NvZGVbIk9rIl0gPSAyMDBdID0gIk9rIjsKICAgIEh0dHBTdGF0dXNDb2RlW0h0dHBTdGF0dXNDb2RlWyJJbUFUZWFwb3QiXSA9IDQxOF0gPSAiSW1BVGVhcG90IjsKfSkoSHR0cFN0YXR1c0NvZGUgfHwgKEh0dHBTdGF0dXNDb2RlID0ge30pKTsKCgpjb25zb2xlLmxvZyhIdHRwU3RhdHVzQ29kZS5JQW1BVGVhcG90KTsAAGZpbGUuanMAaW1wb3J0IHtIdHRwU3RhdHVzQ29kZX0gZnJvbSAnLi9odHRwLmpzJzsKCmNvbnNvbGUubG9nKEh0dHBTdGF0dXNDb2RlLklBbUFUZWFwb3QpOw)

We just discovered that.

**Comment by kbrilla** [comment]
One more question, You can use `enum` as type but it cannot be done with object, also mayby not a big change but if You hover over object You loose info about value 
image
image
also well small TS goodies are lost
image
image
this cannot be done
image

Also adding `as const` does solve some issues and improves on type 
image
but still need to joggle with types to get same functionality
image

So is it some internal change only and to 'public' this will be unchanged or is it breaking change?

**Comment by JeanMeche** [comment]
Yes, this is likely a breaking change, with the PR I wanted to see how breaking it was in Google's codebase !

**Comment by JeanMeche** [comment]
cc @frigus02 since you worked on the removal of the `const enum`.

File: packages/common/http/src/response.ts [hunk 1]
```diff
@@ -358,72 +358,79 @@ export class HttpErrorResponse extends HttpResponseBase implements Error {
...
+  NotExtended: 510,
+  NetworkAuthenticationRequired: 511
+} as const;
+
+/**
+ * Http status codes.
+ * As per https://www.iana.org/assignments/http-status-codes/http-status-codes.xhtml
+ * @publicApi
+ */
+type HttpStatusCode = typeof HttpStatusCode[keyof typeof HttpStatusCode];
```
Comment (kbrilla) [inline packages/common/http/src/response.ts:436]:
```suggestion
export type HttpStatusCode = typeof HttpStatusCode[keyof typeof HttpStatusCode];
```

**Comment by frigus02** [comment]
Hm, I wasn't aware that you typically feed Angular CLI output to esbuild instead of the TS code directly. It makes sense, though. I see that esbuild explicitly calls out that optimizing the enum emit in JS files is not supported (https://esbuild.github.io/api/#minify-considerations), which is unfortunate but also understandable from their side.

This PR seems like a reasonable solution to me. I'm not on the Angular team, though. My opinion doesn't really count 🙂.

**Pull request closed without merging**